Added admin api for managing tags

rrgeorge opened the pull request:
Adding an admin api for managing tags. This adds the `/api/v1/admin/tags` endpoint.
`GET /api/v1/admin/tags` returns a paginated list of all tags
`GET /api/v1/admin/tags/:id` returns a single tag
`PUT /api/v1/admin/tags/:id` updates a single tag using the following attributes:
- `display_name`: set the display name (changing the case)
- `trendable`: boolean: Allow this hashtag to appear under trends
- `usable`: boolean: Allow posts to use this hashtag
- `listable`: boolean: Allow this hashtag to appear in searches and suggestions

rrgeorge committed:
Added admin api for managing tags

ThisIsMissEm commented on spec/controllers/api/v1/admin/tags_controller_spec.rb:
```diff
@@ -0,0 +1,53 @@
+  end
+
+  describe 'GET #index' do
+    it 'returns http success' do
+      get :index
+      expect(response).to have_http_status(200)
+    end
```
Is it worth testing the pagination?

rrgeorge replied:
To be honest, I tried to figure that out and I couldn't get the rspec test to work right.

ThisIsMissEm commented on spec/controllers/api/v1/admin/tags_controller_spec.rb:
```diff
@@ -0,0 +1,53 @@
+  end
+
+  describe 'PUT #update' do
+    let(:scopes) { 'admin:write' }
```
Should we add a scope of `admin:write:taxonomies`, instead of just using the global `admin:write` permission here?

rrgeorge replied:
Is that a new scope? I don't believe it's an option when creating a token.

ClearlyClaire replied:
That is not a scope that exists.

ThisIsMissEm commented on spec/controllers/api/v1/admin/tags_controller_spec.rb:
```diff
@@ -0,0 +1,53 @@
+      expect(tag.reload.display_name).to eq tag.name.upcase
+    end
+
+    it 'returns http unprocessable entity' do
+      put :update, params: { id: tag.id, display_name: tag.name + tag.id.to_s }
+      expect(response).to have_http_status 422
+    end
```
Is this actually testing the TagUpdateErrorSerializer here? I'm not sure if this is a test that covers updating a Tag to be invalid?

rrgeorge replied:
It does

ThisIsMissEm commented on spec/controllers/api/v1/admin/tags_controller_spec.rb:
```diff
@@ -0,0 +1,53 @@
+
+  let(:role)   { UserRole.find_by(name: 'Admin') }
+  let(:user)   { Fabricate(:user, role: role) }
+  let(:scopes)  { 'admin:read admin:write' }
```
Should we have a test for access these APIs without the correct scope?

rrgeorge replied:
I think I can do that.

ThisIsMissEm reviewed (commented):
Just a few comments, mostly related to the test coverage, but looks good!

ClearlyClaire commented on app/controllers/api/v1/admin/tags_controller.rb:
```diff
@@ -0,0 +1,77 @@
+
+class Api::V1::Admin::TagsController < Api::BaseController
+  include Authorization
+  before_action -> { authorize_if_got_token! :'admin:read' }, only: [:index, :show]
+  before_action -> { authorize_if_got_token! :'admin:write' }, only: :update
```
It probably would make sense to have finer permissions, e.g. `admin:read:tags` and `admin:write:tags`, but that can be changed later.

ClearlyClaire commented on app/controllers/api/v1/admin/tags_controller.rb:
```diff
@@ -0,0 +1,77 @@
+    if @tag.update(tag_params.merge(reviewed_at: Time.now.utc))
+      render json: @tag, serializer: REST::Admin::TagSerializer
+    else
+      render json: @tag, serializer: REST::Admin::TagUpdateErrorSerializer, status: 422
```
That seems a bit unnecessary; at least that's not currently common practice in Mastodon.

rrgeorge replied:
We give an error in the web ui, I figured it was worthwhile to return an error on the api too.
I modeled this behavior from the way the admin api for domain blocks works.

ClearlyClaire replied:
IIRC the reason for the admin blocks API doing that is to surface which existing domain block is conflicting so the app can automatically redirect there. I think a simple textual error message is fine here!

ClearlyClaire commented on spec/controllers/api/v1/admin/tags_controller_spec.rb:
This is fine and I won't block on it, but we really want to have requests specs rather than controller specs for these, see `spec/requests/api/v1/admin/`

rrgeorge replied:
I'll rewrite the tests on requests.

ClearlyClaire reviewed (approved):
Looks ok overall, though I'm not completely sure about special-treating the error handling here.

Also, we'd prefer the tests to be request specs rather than controller specs.

rrgeorge committed:
Changed tag update error to just return error message instead of tag object
Added listable to tag serializer
Removed tag error serializer
Removed controller tests
Added requests tests

ClearlyClaire commented on app/controllers/api/v1/admin/tags_controller.rb:
```diff
@@ -0,0 +1,78 @@
+
+  def update
+    authorize @tag, :update?
+    if @tag.update(tag_params.merge(reviewed_at: Time.now.utc))
+      render json: @tag, serializer: REST::Admin::TagSerializer
+    else
+      @tag.reload
+      render json: { error: I18n.t('tags.does_not_match_previous_name') }, status: 422 unless tag_params[:display_name].casecmp(@tag.name.mb_chars).zero?
+    end
```
This won't display anything in the event the reason for the failure is different.

I would just do `@tag.update!(…)` and let the error handling (from `app/controllers/api/base_controller.rb`) catch the exception and return the appropriate error:
```suggestion
    @tag.update!(tag_params.merge(reviewed_at: Time.now.utc))
```

ClearlyClaire reviewed (changes requested):
Thank you for the update!

I would just change the error handling, as per the inline suggestion, otherwise it looks good to me!

rrgeorge committed:
remove ineffective error handling
